add channels last for AdapativeMaxPool2d

mingfeima opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #58348 opitimze ConvTransposedND with mkldnn float32 and bfloat16 on CPU
* #55864 enable BFloat16 mkldnn_convolution on both contiguous and channels last memory format
* #55584 add channels last (2d) support for mkldnn_convolution
* #51185 add channels last support for ConvTranspose2d
* #50573 add channels last support for PixelShuffle and PixelUnshuffle
* #50247 add channels last support for ChannelShuffle
* #49984 add channel last support for MaxUnpool2d
* #49821 add channels last for GroupNorm
* #49582 add channels last support for thnn_conv2d (non-dilated)
* **#48920 add channels last for AdapativeMaxPool2d**

Differential Revision: [D25399467](https://our.internmc.facebook.com/intern/diff/D25399467)

mingfeima committed:
add channels last for AdapativeMaxPool2d

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last for AdapativeMaxPool2d"




[ghstack-poisoned]

mingfeima committed:
Update on "add channels last for AdapativeMaxPool2d"




[ghstack-poisoned]

mingfeima committed:
Update on "add channels last for AdapativeMaxPool2d"




[ghstack-poisoned]

mingfeima committed:
Update on "add channels last for AdapativeMaxPool2d"


Differential Revision: [D25399467](https://our.internmc.facebook.com/intern/diff/D25399467)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last for AdapativeMaxPool2d"


Differential Revision: [D25399467](https://our.internmc.facebook.com/intern/diff/D25399467)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last for AdapativeMaxPool2d"


Differential Revision: [D25399467](https://our.internmc.facebook.com/intern/diff/D25399467)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last for AdapativeMaxPool2d"


Differential Revision: [D25399467](https://our.internmc.facebook.com/intern/diff/D25399467)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last for AdapativeMaxPool2d"


Differential Revision: [D25399467](https://our.internmc.facebook.com/intern/diff/D25399467)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last for AdapativeMaxPool2d"


Differential Revision: [D25399467](https://our.internmc.facebook.com/intern/diff/D25399467)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last for AdapativeMaxPool2d"


Differential Revision: [D25399467](https://our.internmc.facebook.com/intern/diff/D25399467)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last for AdapativeMaxPool2d"


Differential Revision: [D25399467](https://our.internmc.facebook.com/intern/diff/D25399467)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last for AdapativeMaxPool2d"


Differential Revision: [D25399467](https://our.internmc.facebook.com/intern/diff/D25399467)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last for AdapativeMaxPool2d"


Differential Revision: [D25399467](https://our.internmc.facebook.com/intern/diff/D25399467)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last for AdapativeMaxPool2d"


Differential Revision: [D25399467](https://our.internmc.facebook.com/intern/diff/D25399467)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last for AdapativeMaxPool2d"


Differential Revision: [D25399467](https://our.internmc.facebook.com/intern/diff/D25399467)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last for AdapativeMaxPool2d"


Differential Revision: [D25399467](https://our.internmc.facebook.com/intern/diff/D25399467)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last for AdapativeMaxPool2d"


Differential Revision: [D25399467](https://our.internmc.facebook.com/intern/diff/D25399467)

[ghstack-poisoned]

glaringlee commented on aten/src/ATen/native/cpu/AdaptiveMaxPoolKernel.cpp:
```diff
@@ -0,0 +1,341 @@
+  // for the convience of vectorization, use integer of the same size of scalar_t,
+  //   e.g. int32_t for float, int64_t for double
+  // need to make sure doesn't overflow
+  TORCH_CHECK(input_height * input_width < std::numeric_limits<integer_t>::max());
```
I think this check won't work properly if input_height * input_width is bigger than int64_t max since the multiply result might be a small number which will still pass this test. I am not sure whether we will really have such big numbers in reality though, I think probably use `input_height <= std::ceiling(max() / (double)input_width)` is safer.

mingfeima replied:
@glaringlee Updated!
Indeed the check is mostly unnecessary since normally input_height * input_width won't be that big. Add this check only because I need to downcast indexing from int64_t to int32_t (for vectorization purpose).

glaringlee reviewed (commented):
Generally looks good to me, just one comment. Please take a look at.

glaringlee commented on test/test_nn.py:
```diff
@@ -3877,6 +3877,60 @@ def test_adaptive_pooling_avg_nhwc_non_contiguous(self):
             self.assertEqual(out, ref_out)
             self.assertEqual(input.grad, ref_input.grad)
 
+    def test_adaptive_pooling_max_nhwc(self):
```
Can we convert this test into a test case that with device type? so you can use @onlyCPU, just like what you did here:
https://github.com/pytorch/pytorch/pull/49582/files#diff-7927db349f568afca2de9b94d74ea5c3b8cb468cb6a433d0cc1e61e65c515a36R13793

mingfeima replied:
Updated!

mingfeima committed:
Update on "add channels last for AdapativeMaxPool2d"


Differential Revision: [D25399467](https://our.internmc.facebook.com/intern/diff/D25399467)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last for AdapativeMaxPool2d"


Differential Revision: [D25399467](https://our.internmc.facebook.com/intern/diff/D25399467)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last for AdapativeMaxPool2d"


Differential Revision: [D25399467](https://our.internmc.facebook.com/intern/diff/D25399467)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last for AdapativeMaxPool2d"


Differential Revision: [D25399467](https://our.internmc.facebook.com/intern/diff/D25399467)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last for AdapativeMaxPool2d"


Differential Revision: [D25399467](https://our.internmc.facebook.com/intern/diff/D25399467)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last for AdapativeMaxPool2d"


Differential Revision: [D25399467](https://our.internmc.facebook.com/intern/diff/D25399467)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last for AdapativeMaxPool2d"


Differential Revision: [D25399467](https://our.internmc.facebook.com/intern/diff/D25399467)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last for AdapativeMaxPool2d"


Differential Revision: [D25399467](https://our.internmc.facebook.com/intern/diff/D25399467)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last for AdapativeMaxPool2d"


Differential Revision: [D25399467](https://our.internmc.facebook.com/intern/diff/D25399467)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last for AdapativeMaxPool2d"


Differential Revision: [D25399467](https://our.internmc.facebook.com/intern/diff/D25399467)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last for AdapativeMaxPool2d"


Differential Revision: [D25399467](https://our.internmc.facebook.com/intern/diff/D25399467)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last for AdapativeMaxPool2d"


Differential Revision: [D25399467](https://our.internmc.facebook.com/intern/diff/D25399467)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last for AdapativeMaxPool2d"


Differential Revision: [D25399467](https://our.internmc.facebook.com/intern/diff/D25399467)

[ghstack-poisoned]

ngimel commented on aten/src/ATen/native/cpu/AdaptiveMaxPoolKernel.cpp:
```diff
@@ -0,0 +1,341 @@
+  // for the convience of vectorization, use integer of the same size of scalar_t,
+  //   e.g. int32_t for float, int64_t for double
+  // need to make sure doesn't overflow
+  TORCH_CHECK(input_height <= std::ceil(std::numeric_limits<integer_t>::max() / (double)input_width));
```
Please don't use conversions here, if you want to make sure that integer_t is enough to address spatial dimensions, do 
```
TORCH_CHECK (input_height * input_width <= std::numeric_limits<integer_t>::max() )
```

mingfeima replied:
glaringlee previously had a review at this point here:
https://github.com/pytorch/pytorch/pull/48920#discussion_r594436426
I think he is right, if the spatial dimension overflows int32, the multiplied result would be a small number.

anyway, it is unlikely that input_height * input_width would overflow int32.

ngimel replied:
It's not only unlikely, a tensor with more than int64_t elements would need 4e10 GB memory, which is many orders of magnitude more than the biggest systems.

mingfeima replied:
OK I see, updated!

mingfeima committed:
Update on "add channels last for AdapativeMaxPool2d"


Differential Revision: [D25399467](https://our.internmc.facebook.com/intern/diff/D25399467)

[ghstack-poisoned]

ngimel commented on aten/src/ATen/native/cpu/AdaptiveMaxPoolKernel.cpp:
```diff
@@ -0,0 +1,341 @@
+    const Tensor& grad_input,
+    const Tensor& grad_output,
+    const Tensor& indices) {
+  switch (grad_output.suggest_memory_format()) {
```
this won't work properly, in many cases adaptiveMaxPooling has an output size of (1,1), and suggested memory format won't be ChannelsLast. You'd need to pass `input` as argument here, to get suggested memory format from `input`.

mingfeima replied:
right! I updated with "grad_input.suggest_memory_format()"

grad_input is created via:
```C++
set_output(0, input.sizes(), input.options().memory_format(input.suggest_memory_format()));
```
has the same memory format with input.

ngimel commented on test/test_nn.py:
```diff
@@ -13368,6 +13368,58 @@ def helper(n, c, h, w, ks):
+    @onlyCPU
+    @dtypes(torch.float, torch.double)
+    def test_adaptive_pooling_max_nhwc(self, device, dtype):
+        input = torch.randint(1, 10, (4, 8, 8, 8), dtype=dtype).to(device)
```
`device=device`

mingfeima replied:
updated!

ngimel commented on test/test_nn.py:
```diff
@@ -13368,6 +13368,58 @@ def helper(n, c, h, w, ks):
+    @onlyCPU
+    @dtypes(torch.float, torch.double)
+    def test_adaptive_pooling_max_nhwc(self, device, dtype):
+        input = torch.randint(1, 10, (4, 8, 8, 8), dtype=dtype).to(device)
```
please test input with non-square width and height, to avoid accidental bugs where sizes are swapped in kernels

mingfeima replied:
add non square width and height as well as output_height and output_width

ngimel commented on test/test_nn.py:
```diff
@@ -13368,6 +13368,58 @@ def helper(n, c, h, w, ks):
+        input = torch.randint(1, 10, (4, 8, 8, 8), dtype=dtype).to(device)
+        input = input.contiguous(memory_format=torch.channels_last).requires_grad_()
+        grad = torch.randint(1, 10, (4, 8, 7, 7), dtype=dtype).to(device)
+        pool = torch.nn.AdaptiveMaxPool2d((7, 7), return_indices=True).to(device)
```
same here, non-square output size. If you fix 1x1 case, please test with 1x1 gradOutput also.

mingfeima replied:
updated!

ngimel commented on test/test_nn.py:
```diff
@@ -13368,6 +13368,58 @@ def helper(n, c, h, w, ks):
+    def test_adaptive_pooling_max_nhwc(self, device, dtype):
+        input = torch.randint(1, 10, (4, 8, 8, 8), dtype=dtype).to(device)
+        input = input.contiguous(memory_format=torch.channels_last).requires_grad_()
+        grad = torch.randint(1, 10, (4, 8, 7, 7), dtype=dtype).to(device)
```
`grad` is not channels-last, so you are not testing channels-last backward path.

mingfeima replied:
fixed!

ngimel commented on test/test_nn.py:
```diff
@@ -13368,6 +13368,58 @@ def helper(n, c, h, w, ks):
+        grad = grad[:, ::2, :, :]
+        pool = torch.nn.AdaptiveMaxPool2d((7, 7), return_indices=True).to(device)
+
+        ref_input = input.detach().clone().contiguous().requires_grad_(True)
```
both these tests can use the same helper function, with the different inputs.

mingfeima replied:
updated!

mingfeima committed:
Update on "add channels last for AdapativeMaxPool2d"


Differential Revision: [D25399467](https://our.internmc.facebook.com/intern/diff/D25399467)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last for AdapativeMaxPool2d"


Differential Revision: [D25399467](https://our.internmc.facebook.com/intern/diff/D25399467)

[ghstack-poisoned]

mingfeima commented on aten/src/ATen/native/cpu/AdaptiveMaxPoolKernel.cpp:
```diff
@@ -0,0 +1,341 @@
+  }
+}
+
+void adapative_max_pool2d_backward_kernel_impl(
```
updated!

mingfeima commented:
@ngimel thanks for the review! I have updated this patch, please check!

mingfeima committed:
Update on "add channels last for AdapativeMaxPool2d"


Differential Revision: [D25399467](https://our.internmc.facebook.com/intern/diff/D25399467)

[ghstack-poisoned]

mingfeima committed:
Update on "add channels last for AdapativeMaxPool2d"


Differential Revision: [D25399467](https://our.internmc.facebook.com/intern/diff/D25399467)

[ghstack-poisoned]

mingfeima commented:
rebased!

VitalyFedyunin commented:
Could you please rebase once more as I landed 2 older and related PRs.

mingfeima committed:
Update on "add channels last for AdapativeMaxPool2d"


Differential Revision: [D25399467](https://our.internmc.facebook.com/intern/diff/D25399467)

[ghstack-poisoned]

mingfeima commented:
@VitalyFedyunin, rebased to lastest master! please check

VitalyFedyunin commented:
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D25399467).
